examples/schelling_segregation/src/main.cu

```diff
@@ -0,0 +1,357 @@
+#include <iostream>
+#include <cstdio>
+#include <cstdlib>
+#include <fstream>
+#include <array>
+
+#include "flamegpu/flamegpu.h"
+
+// Configurable properties
+constexpr unsigned int GRID_WIDTH = 100;
+constexpr float THRESHOLD = 0.70;
+
+constexpr unsigned int A = 0;
+constexpr unsigned int B = 1;
+constexpr unsigned int UNOCCUPIED = 2;
+
+// Agents output their type
+FLAMEGPU_AGENT_FUNCTION(output_type, flamegpu::MessageNone, flamegpu::MessageArray2D) {
+    FLAMEGPU->message_out.setVariable<unsigned int>("type", FLAMEGPU->getVariable<unsigned int>("type"));
+    FLAMEGPU->message_out.setIndex(FLAMEGPU->getVariable<unsigned int, 2>("pos", 0), FLAMEGPU->getVariable<unsigned int, 2>("pos", 1));
+    return flamegpu::ALIVE;
+}
+
+// Agents decide whether they are happy or not and whether or not their space is available
+FLAMEGPU_AGENT_FUNCTION(determine_status, flamegpu::MessageArray2D, flamegpu::MessageNone) {
+    const unsigned int my_x = FLAMEGPU->getVariable<unsigned int, 2>("pos", 0);
+    const unsigned int my_y = FLAMEGPU->getVariable<unsigned int, 2>("pos", 1);
+
+    unsigned int same_type_neighbours = 0;
+    unsigned int diff_type_neighbours = 0;
+
+    // Iterate 3x3 Moore neighbourhood (this does not include the central cell)
+    unsigned int my_type = FLAMEGPU->getVariable<unsigned int>("type");
+    for (auto &message : FLAMEGPU->message_in.wrap(my_x, my_y)) {
+        int message_type = message.getVariable<unsigned int>("type");
+        same_type_neighbours += my_type == message_type;
+        diff_type_neighbours += (my_type != message_type) && (message_type != UNOCCUPIED);
+    }
+
+    int isHappy = (static_cast<float>(same_type_neighbours) / (same_type_neighbours + diff_type_neighbours)) > THRESHOLD;
+    FLAMEGPU->setVariable<unsigned int>("happy", isHappy);
+    unsigned int my_next_type = ((my_type != UNOCCUPIED) && isHappy) ? my_type : UNOCCUPIED;
+    FLAMEGPU->setVariable<unsigned int>("next_type", my_next_type);
+    FLAMEGPU->setVariable<unsigned int>("movement_resolved", (my_type == UNOCCUPIED) || isHappy);
+    unsigned int my_availability = (my_type == UNOCCUPIED) || (isHappy == 0);
+    FLAMEGPU->setVariable<unsigned int>("available", my_availability);
+
+    return flamegpu::ALIVE;
+}
+
+FLAMEGPU_AGENT_FUNCTION_CONDITION(is_available) {
```

Comment:
Might benefit from a comment above this, to denote the remaining functions are from the iterative submodel.

(My initial thought was to split the entire submodel into a seperate header or source file, like I've done in primage to aid readability, but probably overkill for the small model (and I doubt I did it for sugar scape).